lib/astarte_vmq_plugin/connection/synchronizer.ex

```diff
@@ -0,0 +1,171 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2022 SECO Mind Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.VMQ.Plugin.Connection.Synchronizer do
+  @moduledoc """
+    This module handles connection and disconnection events. As it may happen that
+    VerneMQ hooks are called in the wrong order if events occur in a delta of ~10 ms,
+    the module forces the correct order of publication (i.e. disconnection before reconnection).
+    See https://github.com/vernemq/vernemq/issues/1741.
+  """
+  alias Astarte.VMQ.Plugin
+  @behaviour :gen_statem
+
+  # Hooks might be called in the wrong order if the time difference is ~10 ms.
+  # Let's play it safe with 50 ms.
+  @timeout_ms 50
+
+  defmodule Data do
+    defstruct [
+      :client_id,
+      :connection_headers,
+      :connection_timestamp,
+      :disconnection_headers,
+      :disconnection_timestamp
+    ]
+  end
+
+  # API
+
+  @impl true
+  def callback_mode() do
+    [:state_functions]
+  end
+
+  def child_spec(opts) do
+    %{
+      id: __MODULE__,
+      start: {__MODULE__, :start_link, [opts]},
+      type: :worker,
+      restart: :transient
+    }
+  end
+
+  def start_link(opts) do
+    client_id = Keyword.fetch!(opts, :client_id)
+    name = via_tuple(client_id)
+
+    with {:ok, pid} <- :gen_statem.start_link(name, __MODULE__, opts, []) do
+      {:ok, pid}
+    end
+  end
+
+  def handle_connection(pid, timestamp, additional_headers \\ []) do
+    :gen_statem.cast(pid, {:connection, timestamp, additional_headers})
```

Comment:
Now that I see it - are we sure we want to do this with a `cast` instead of a `call`? `cast`s are all good and well, except when they're not (e.g. they don't provide backpressure, they don't ensure the callee has correctly handled the message etc). I'd probably switch both this and the function below to `call`.

Reply:
Another bonus: using `call` you (probably) can remove some sleeps in tests

Reply:
After discussing about this IRL, `cast` is actually the right choice to avoid some corner cases that `call` would pose

Reply:
ackchually